### PR TITLE
8263375: Support stack watermarks in Zero VM

### PR DESCRIPTION
**Pull request by shipilev**
Zero VM supports most of GCs. Since JDK 16, Shenandoah uses stack watermarks, so Zero has to support those if Shenandoah+Zero support is to remain. This PR adds the stack watermark support in Zero VM. This should also be useful as other projects, notably Loom, mature and depend on stack watermarks.

Zero already calls into Hotspot safepoint machinery to do things, and it seems only the hooks for `on_iteration` and `on_unwind` are missing. AFAICS, Zero only has on-return safepoints, renamed it to be more precise.

@fisk, do you see any obvious problems with this patch?

Additional testing:
 - [x] Linux x86_64 Zero `hotspot_gc_shenandoah` now passes

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263375](https://bugs.openjdk.java.net/browse/JDK-8263375): Support stack watermarks in Zero VM

### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4728/head:pull/4728` \
`$ git checkout pull/4728`

Update a local copy of the PR: \
`$ git checkout pull/4728` \
`$ git pull https://git.openjdk.java.net/jdk pull/4728/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4728`

View PR using the GUI difftool: \
`$ git pr show -t 4728`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4728.diff">https://git.openjdk.java.net/jdk/pull/4728.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/zero/zeroInterpreter_zero.cpp [hunk 1]
```diff
@@ -200,6 +200,10 @@ void ZeroInterpreter::main_loop(int recurse, TRAPS) {
     }
     fixup_after_potential_safepoint();
 
+    // Notify the stack watermarks machinery that we are unwinding.
+    // Should do this before resetting the frame anchor.
+    stack_watermark_unwind_check(thread);
```
Comment (fisk) [inline src/hotspot/cpu/zero/zeroInterpreter_zero.cpp:205]: I wonder if this should maybe move down a bit to where we inspect the reason we left the interpreter loop. There are multiple reasons and only some involve unwinding. I'm thinking BytecodeInterpreter::return_from_method and BytecodeInterpreter::do_osr

Regarding BytecodeInterpreter::throwing_exception the current contract for exception handing is that an unwind handler is called *after* unwinding instead. We have some exception handler function in the interpreter runtime that gets called after unwinding with an exception into an interpreted frame. Hopefully that still gets called when using zero. Worth double checking.

**Review by fisk — COMMENTED** [review]
Looks good in general, would just like slightly more precise hooks as explained in a comment above. It will probably work without that though.

File: src/hotspot/cpu/zero/zeroInterpreter_zero.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/cpu/zero/zeroInterpreter_zero.cpp:205]: Thanks! I'll take a look after I am back from extended time off. 

Zero does not actually do OSR anymore (Zero gradually eroded to interpreter-only mode), so only return_from_method might need handling. I'll see what happens in throwing_exception case; worst case I think I can set up a one-off frame anchor and call the unwind handler with it.

Comment (shipilev) [reply src/hotspot/cpu/zero/zeroInterpreter_zero.cpp:205]: OK, I think for `throwing_exception` case you mean `InterpreterRuntime::exception_handler_for_exception` that is called to do `StackWatermarkSet::after_unwind`. That thing is [called](https://github.com/openjdk/jdk/blob/0e14bf70cf6e482a2ec7d13ed37df0bee911740d/src/hotspot/share/interpreter/zero/bytecodeInterpreter.cpp#L2492) from Zero during exception throwing, similar to the places in `TemplateInterpreter`-s for other arches.

New commit now only calls stack watermark check for `return_from_method` and `do_osr`.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by shipilev** [comment]
I suppose there are no other interested parties for this kind of patch, so I'll integrate this soon.

**Pull request closed without merging**